### PR TITLE
RUST-735 Remove Document as default generic type on Collection and Cursor

### PR DESCRIPTION
**Pull request by patrickfreed**
RUST-735

This PR removes `Document` as the default generic type for `Collection` and `Cursor` and updates `Database::collection` and `Database::collection_with_options` to return `Collection<T>`. As part of this, the existing typed helpers became redundant and were removed. These changes were made to ease and promote the use of serde models with the driver instead of regular Documents.

Note: I didn't update the README examples to avoid confusing users of the 1.2.x or 2.0.0-alpha versions. I filed RUST-736 to cover the work for updating that, which we can do right before the release.

### CODE REVIEW

File: src/db/mod.rs
Comment (patrickfreed) [inline src/db/mod.rs:138]: I know we discussed the possibility of introducing a `collection_with_document` helper (or something like that) as an ergonomic way to retrieve a `Collection<Document>`, but after writing a few of these conversions, it seemed simpler / easier to just use the turbofish (`::<>`) operator on the existing helpers.

e.g.
```rust
let collection: Collection<Document> = db.collection("foo"); // explicit type way
let collection = db.collection::<Document>("foo"); // turbofish
let collection = db.collection_with_document("foo"); // helper
```

To me, the helper didn't really make it any easier to type nor did it actually make it clearer what was happening, so it seemed fine to me to just not have it. Thoughts?

File: src/db/mod.rs
Comment (patrickfreed) [inline src/db/mod.rs:197]: In Swift, we have this return a `MongoCursor<CollectionSpecification>`, and I think it would be a nice improvement here as well for 2.0. The versioned API will ensure that we won't have to worry about deserialization errors for the foreseeable future.

File: src/sync/coll.rs
Comment (patrickfreed) [inline src/sync/coll.rs:156]: We could consider returning a `Cursor<O>` here (where `O` is a generic parameter to the `aggregate` method, not the `T` from the `Collection`). Or should we have a separate `aggregate_with_output_type` or something? I think with aggregation it's more likely that a user might just want to retrieve a document than with a `find`, since they may mutate the output along the way, so it's a little riskier to have this one be pure generic. Thoughts?

**Review by patrickfreed — COMMENTED** [review]
I have a few non-as-yet-discussed comments/questions about the public API, so tagging in @kmahar as well.

File: src/cursor/mod.rs
Comment (kmahar) [inline src/cursor/mod.rs:142]: this is honestly not really relevant to the PR, but I'm just curious for my own knowledge and future reference for Swift- aren't some cursors not actually tied to particular collections e.g. one resulting from db-level aggregate? how does this work in that case?

File: src/db/mod.rs
Comment (kmahar) [reply src/db/mod.rs:197]: sgtm (though that's probably not surprising since I already agreed to it once...) 

if so, would be nice to use a `DatabaseSpecification` type for `list_databases` too (assuming `listDatabases` is also in API version 1?)

File: src/db/mod.rs
Comment (kmahar) [reply src/db/mod.rs:138]: from my outsider perspective it doesn't seem to make a significant difference, but I'd defer to you two on whether `collection_with_document` is in any way more idiomatic

File: src/sync/coll.rs
Comment (kmahar) [reply src/sync/coll.rs:156]: per slack conversation: seems we're all on the same page the `T` from the collection should not be used since people often do transformations here.

as for whether to require a type be specified in this method (which I was not thinking about the option of in the slack discussion), or just having a separate `aggregate_with_type` I don't have particularly strong feelings. I agree with your comment that it does seem a bit at odds with the choice to make requiring a type the default for `db.collection`. OTOH maybe it's less likely that people will bother writing custom `Deserialize` types for the output types of single methods?

File: src/db/mod.rs
Comment (patrickfreed) [reply src/db/mod.rs:197]: Good call, I filed RUST-740 to cover the both of them.

File: src/sync/coll.rs
Comment (patrickfreed) [reply src/sync/coll.rs:156]: Yeah, my original inclination was to just leave this as `Document` by default and introduce the new separate method, since I imagine it'll be pretty common to use `Document` at least while prototyping the pipeline. I'm starting to flip back to wanting to maintain consistency though, given how easy it is to just append a `::<Document>` to the end of the `aggregate` call. Especially in the face of complex pipelines, handling the generic type would be a piece of cake.

_However_

This got me trying it out and I noticed an interesting / frustrating technical challenge to doing it the `Cursor<O>` way: turbofish cannot be used in conjunction with `impl Trait` in the argument position (yet, see https://github.com/rust-lang/rust/issues/83701).

e.g.
```rust
fn foo<T>(bar: impl ToString) { ... }
foo::<usize>("blah"); // compile error!
```
yields the following compiler error:
```
error[E0632]: cannot provide explicit generic arguments when `impl Trait` is used in argument position
  --> src/main.rs:21:11
   |
21 |     foo::<usize>("hello");
   |           ^^^^^ explicit generic argument not allowed

```
So you'd have to do it the explicit type parameter way:
```rust
let x: Cursor<Document> = coll.aggregate(vec![], None).await?
```
Which could be pretty frustrating.

This will make it tough to implement the one that takes in a generic argument, since we use impl Trait in all our CRUD methods for our options types (e.g. to be able to say `options` instead of having to say `Some(options)`). This makes me think we may want to do the builder approach here and have the method be on `Cursor` actually, something like:

```rust
let cursor = coll.aggregate(vec![ ... ], None).await?.with_type::<MyDeserializeType>();
```
This is pretty ugly admittedly, but I think it's preferable to breaking consistency with how options are provided to the other CRUD methods or _requiring_ users to use explicit type annotations instead of turbofish.

Comment (kmahar) [reply src/sync/coll.rs:156]: I agree changing the way we pass options either here or everywhere doesn't seem like a great path forward.

is turbofish common/used rather than type annotations that people would be running into this a lot?

adding the method to cursor seems reasonable though it does look clunky.... it also seems consistent with the `clone_with_type` behavior that we currently have on `Collection` to change the generic type. and maybe it would be useful in other use cases as well - e.g. if you provide a `projection` via `FindOptions` that changes the shape of your documents as well, and you could call this on the resulting cursor.

Comment (isabelatkinson) [reply src/sync/coll.rs:156]: Yeah none of these options is very ideal to me; `with_type` seems like it would be more ergonomic for chaining method calls together, since providing a type annotation would require breaking out the call to `aggregate` into a separate line. Updating our CRUD API to use the builder pattern would get rid of this issue, right?

In any case since we're opting to go the `aggregate_with_type` route, I think we can defer coming to a conclusion here til we implement that (and make a decision regarding the builder pattern).

**Review by isabelatkinson — COMMENTED** [review]
Code looks good, just a few questions. Also wanted to double check that the examples in the README are still good, I believe they should be unaffected but would be worth taking a glance to sanity check if you haven't yet. (It would also probably be a good idea to put those in a test somewhere to make sure they don't become outdated.)

File: src/sync/coll.rs
Comment (patrickfreed) [reply src/sync/coll.rs:156]: I think even with the builder pattern for options/sessions we'd still have the same issue unfortunately, since we use `impl IntoIterator` for the pipeline.

Also, to clarify, I don't think `aggregate_with_type` method will really be possible, since we'd have to ditch the `impl Trait` arguments, so I think we kind of have to do the `Cursor::with_type` approach. That being said, we don't need to implement that functionality now.

**Review by patrickfreed — COMMENTED** [review]
I was originally planning on holding off on updating the README until close to the release (RUST-736), but now that we're planning on doing a second alpha I went ahead and updated them. We can still use RUST-736 to track the work for updating the examples to be use serde more prominently. Also, good point on having tests for them--I added a file to our `tests` folder since the examples only use the public API to ensure the examples compile.

File: src/sync/coll.rs
Comment (isabelatkinson) [reply src/sync/coll.rs:156]: Ah right, I forgot we use `impl` for the pipeline.

**Review by isabelatkinson — APPROVED** [review]

**Comment by isabelatkinson** [comment]
whoops, just noticed there's a clippy failure in the readme file (lgtm otherwise though, no need for another round of review on taht)

**Review by kmahar — APPROVED** [review]

**Comment by seanpianka** [comment]
After updating from alpha to alpha.1, I'm now seeing this run-time error with the same code:
```
failed to update collection: Error { kind: ArgumentError { message: \"update document must have first key starting with \\\'$\" }, labels: [] }" }
```

**Comment by patrickfreed** [comment]
The `update_` family of methods are meant to be used with [update operator expressions](https://docs.mongodb.com/manual/reference/command/update/#std-label-update-command-upsert) rather than replacement documents, which the `replace` family of methods are mean to be used with, but we only started validating this in alpha.1 (RUST-718). To fix your issue, update any usages of `update_one` that use a replacement document to use `replace_one` instead.

e.g.
```rust
coll.update_one(doc! { "x": 2 }, doc! { "x": 3 }, None).await?; // error
coll.replace_one(doc! { "x": 2 }, doc! { "$inc": { "x": 1 } }, None).await?; // error

coll.update_one(doc! { "x": 2 }, doc! { "$inc": { "x": 1 } }, None).await?; // correct
coll.replace_one(doc! { "x": 2 }, doc! { "x": 3 }, None).await?; // correct
```

If you have any more questions or concerns, please [open a ticket on our Jira project](https://github.com/mongodb/mongo-rust-driver#bug-reporting--feature-requests) rather than continuing the discussion here or on any other PR, since it can be difficult for the team or other users to find the discussion in the future otherwise.

**Comment by seanpianka** [comment]
Gotcha, I appreciate your detailed reply, thank you @patrickfreed! In the future, I'll open a ticket in the Jira project. 

I managed to fix the `update_one` call via `doc! { "$set": replacement_doc }`, but `replace_one` more closely captures my intent. I'll out that, thanks again!

**Comment by patrickfreed** [comment]
No problem, glad you were able to get it working!

**Pull request merged**
